File: client/src/api/excelImport.ts
import { apiClient } from "./apiClient";

/**
 * Excel import API. Mirrors app/excel_router.py.
 *
 * One shared parse endpoint turns any input-sheet variant into the form
 * suggestions plus the raw workbook; the parsed workbook travels back
 * inside the commit body so the browser never uploads file bytes twice.
 */

export interface ProjectInfoData {
  project_type: string | null;
  project_location: string | null;
  client_name: string | null;
  borehole_number: string | null;
  project_id: string | null;
  start_date: string | null;
  end_date: string | null;
  final_depth: number | null;
}

export interface BorelogRow {
  depth_from: number;
  depth_to: number;
  soil_description: string | null;
  sand_clay: string | null;
}

export interface RockProfileRow {
  depth_from: number;
  depth_to: number;
  rock_description: string | null;
  recovery: number | null;
  rqd: number | null;
  remark: string | null;
}

export interface SPTImportRow {
  spt_depth: number;
  blows_15: number | null;
  blows_30: number | null;
  blows_45: number | null;
  n_value: number | null;
}

export interface ImportedWorkbook {
  project_info: ProjectInfoData;
  borelog: BorelogRow[];
  rock_profile: RockProfileRow[];
  spt: SPTImportRow[];
}

export interface ImportSuggestions {
  project_name: string | null;
  project_location: string | null;
  project_client: string | null;
  borehole_id: string | null;
  start_date: string | null;
  end_date: string | null;
  final_depth: number | null;
}

export interface ParsedWorkbookResponse {
  workbook: ImportedWorkbook;
  suggestions: ImportSuggestions;
}

export async function parseExcelWorkbook(
  file: File
): Promise<ParsedWorkbookResponse> {
  const body = new FormData();
  body.append("file", file);

  const response = await apiClient("/api/excel/workbook", {
    method: "POST",
    body,
  });

  if (!response.ok) {
    const errorBody = await response.json().catch(() => null);
    throw new Error(errorBody?.detail ?? "Unable to read this workbook.");
  }

  return response.json();
}

export interface ProjectExcelImportPayload {
  project_id: string;
  project_name: string;
  project_location?: string | null;
  project_client?: string | null;
  borehole_id?: string | null;
  borehole_type: string;
  start_date?: string | null;
  end_date?: string | null;
  final_depth?: number | null;
  workbook: ImportedWorkbook;
}

export interface ImportResult {
  message: string;
  project_id: string;
  borehole_id: string;
  geol_rows: number;
  core_rows: number;
  spt_rows: number;
}

export async function createProjectFromExcel(
  payload: ProjectExcelImportPayload
): Promise<ImportResult> {
  const response = await apiClient("/api/excel/projects", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });

  if (!response.ok) {
    const errorBody = await response.json().catch(() => null);
    throw new Error(
      errorBody?.detail ?? "Unable to import this workbook."
    );
  }

  return response.json();
}

export async function createLocaFromExcel(
  projectId: string,
  payload: { borehole_type: string; workbook: ImportedWorkbook }
): Promise<ImportResult> {
  const response = await apiClient(
    `/api/excel/projects/${encodeURIComponent(projectId)}/locas`,
    {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(payload),
    }
  );

  if (!response.ok) {
    const errorBody = await response.json().catch(() => null);
    throw new Error(
      errorBody?.detail ?? "Unable to import this workbook."
    );
  }

  return response.json();
}

File: server/app/schemas/excel_import.py
"""
Excel import schemas.

The workbook shape produced by app.excel_parser.parse_input_sheet and the
bodies for the two import endpoints that commit it:

  ProjectExcelImport -> "Add New Project" page (creates PROJ + first LOCA)
  LocaExcelImport    -> Overview "import" (adds a LOCA to an existing project)

Field mapping is fixed by the existing AGS data contract (see
schemas/field_data.py); importing introduces no new database columns:

  Borelog row  -> GEOL (GEOL_TOP, GEOL_BASE, GEOL_DESC, GEOL_GEOL)
  Rock row     -> CORE (CORE_TOP, CORE_BASE, CORE_PREC, CORE_RQD, CORE_REM)
  SPT row      -> ISPT (ISPT_TOP, ISPT_INC1..INC3, ISPT_NVAL)
  Project info -> PROJ (PROJ_LOC, PROJ_CLNT) + LOCA (LOCA_ID, LOCA_STAR,
                  LOCA_ENDD, LOCA_FDEP, LOCA_TYPE)
"""

from datetime import date
from decimal import Decimal
from typing import List, Optional

from pydantic import BaseModel, Field


class ProjectInfoData(BaseModel):
    project_type: Optional[str] = None
    project_location: Optional[str] = None
    client_name: Optional[str] = None
    borehole_number: Optional[str] = None
    # Sheets do not carry it today, but a project-scoped export may state
    # which project it belongs to; the loca import verifies it on commit.
    project_id: Optional[str] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    # Decimal, not float: JSON numbers are parsed to their exact decimal
    # digits, so 5.8 stays 5.8 through the round trip and asyncpg stores
    # exactly that in NUMERIC columns (a float would land as
    # 5.79999999999999982236431605997495353221893310546875).
    final_depth: Optional[Decimal] = None


class BorelogRow(BaseModel):
    depth_from: Decimal
    depth_to: Decimal
    soil_description: Optional[str] = None
    sand_clay: Optional[str] = None


class RockProfileRow(BaseModel):
    depth_from: Decimal
    depth_to: Decimal
    rock_description: Optional[str] = None
    recovery: Optional[Decimal] = None
    rqd: Optional[Decimal] = None
    remark: Optional[str] = None


class SPTImportRow(BaseModel):
    spt_depth: Decimal
    blows_15: Optional[Decimal] = None
    blows_30: Optional[Decimal] = None
    blows_45: Optional[Decimal] = None
    n_value: Optional[Decimal] = None


class ImportedWorkbook(BaseModel):
    project_info: ProjectInfoData = Field(default_factory=ProjectInfoData)
    borelog: List[BorelogRow] = Field(default_factory=list)
    rock_profile: List[RockProfileRow] = Field(default_factory=list)
    spt: List[SPTImportRow] = Field(default_factory=list)


class ProjectExcelImport(BaseModel):
    """Commit body for the Add New Project page.

    Form values override the sheet when present, so a user can correct a
    prefilled field before submitting; anything left blank falls back to
    the workbook's ProjectInfo sheet.
    """

    project_id: str = Field(min_length=1)
    project_name: str = Field(min_length=1)

    project_location: Optional[str] = None
    project_client: Optional[str] = None

    borehole_id: Optional[str] = None
    borehole_type: str = "BH"
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    final_depth: Optional[Decimal] = None

    workbook: ImportedWorkbook


class LocaExcelImport(BaseModel):
    """Commit body for the Overview import (location into an open project).

    The location identity comes from the workbook itself; only the type,
    which the sheet does not carry, is supplied here.
    """

    borehole_type: str = "BH"
    workbook: ImportedWorkbook
